### DISCUSSION TITLE
Selecting Specific Time Steps for Rainfall Data Visualization

### DISCUSSION THREAD

**Discussion by haritha1022** [pydata/xarray#6261] [Q&A] [answered]
### What happened?

i just want plot particular time steps variable data
define rain=((rainnc(t=17))-rainnc(t=9)))
how can do it in python
any idea

### What did you expect to happen?

i want data only some time steps 
how can i eleminate other timesteps

### Minimal Complete Verifiable Example

```python
import xarray as xr
import cartopy.crs as ccrs
from cartopy import feature as cf
import matplotlib.pyplot as plt
import numpy as np

#Opening and reading data
data=xr.open_dataset("heavyrain3.nc")
print(data)


lon=data.lon[:]
lat=data.lat[:]


#lev_2=data.lev_2[:]

rainnc=data.rainnc[17-9]
#time=data.time[0]







ax = plt.axes(projection=ccrs.PlateCarree())

rainnc.isel(time=16, lev_2=0).plot.pcolormesh(cmap='Purples',add_colorbar=False);


#ax.set_extent([west, east, south, north])
ax.coastlines()
ax.add_feature(cf.BORDERS)

plt.savefig("rain31.jpg")
plt.show()
```

### Relevant log output

```python
raise ValueError(
ValueError: Dimensions {'time'} do not exist. Expected one or more of ('lev_2', 'lat', 'lon')
```

### Environment

Traceback (most recent call last):
  File "/home/user/Documents/IGCAR/rain/rain.py", line 39, in <module>
    rainnc.isel(time=16, lev_2=0).plot.pcolormesh(cmap='Purples',add_colorbar=False);
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/dataarray.py", line 1208, in isel
    variable = self._variable.isel(indexers, missing_dims=missing_dims)
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/variable.py", line 1190, in isel
    indexers = drop_dims_from_indexers(indexers, self.dims, missing_dims)
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/utils.py", line 834, in drop_dims_from_indexers
    raise ValueError(
ValueError: Dimensions {'time'} do not exist. Expected one or more of ('lev_2', 'lat', 'lon')

**Comment by max-sixty** [comment]
Please review the links on producing quality examples. I'll close for now but feel free to reopen if you're up for producing one.

**Comment by haritha1022** [comment]
```python
import xarray as xr
import cartopy.crs as ccrs
from cartopy import feature as cf
import matplotlib.pyplot as plt
import numpy as np

data=xr.open_dataset("rainncreq.nc")
print(data)
lon=data.lon[:]
lat=data.lat[:]
lev_2=data.lev_2[0]
rainnc=data.rainnc[:]
ax = plt.axes(projection=ccrs.PlateCarree())
rainnc.isel(time=0+8,lev_2=0).plot.pcolormesh(cmap='jet',vmax=100,vmin=0,add_colorbar=True);
ax.coastlines()
ax.add_feature(cf.BORDERS)
plt.savefig("rain31.jpg")
plt.show()
```
then it is not giving error just the data
but actually i want add the the time steps from 0 to 8
when i was 0+1+2+3....+8
then it is giving index error

```pytb
File "/home/user/Documents/IGCAR/rain/rain.py", line 36, in <module>
    rainnc.isel(time=0+1+2+3+4+5+6+7+8,lev_2=0).plot.pcolormesh(cmap='jet',vmax=100,vmin=0,add_colorbar=True);
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/dataarray.py", line 1208, in isel
    variable = self._variable.isel(indexers, missing_dims=missing_dims)
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/variable.py", line 1193, in isel
    return self[key]
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/variable.py", line 786, in __getitem__
    data = as_indexable(self._data)[indexer]
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/indexing.py", line 555, in __getitem__
    return type(self)(_wrap_numpy_scalars(self.array[key]))
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/indexing.py", line 524, in __getitem__
    return type(self)(_wrap_numpy_scalars(self.array[key]))
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/indexing.py", line 431, in __getitem__
    return type(self)(self.array, self._updated_key(indexer))
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/indexing.py", line 403, in _updated_key
    full_key.append(_index_indexer_1d(k, next(iter_new_key), size))
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/core/indexing.py", line 176, in _index_indexer_1d
    indexer = _expand_slice(old_indexer, size)[applied_indexer]
IndexError: index 36 is out of bounds for axis 0 with size 9
```

hope you understand my problem

**Reply by dcherian** [reply]
Please see https://docs.xarray.dev/en/stable/user-guide/indexing.html#indexing-with-dimension-names particularly the examples using `slice`. This syntax is similar to standard python syntax for lists, strings, and tuples. For example,see https://towardsdatascience.com/mastering-indexing-and-slicing-in-python-443e23457125

**Comment by haritha1022** [comment]
good morning sir
   slice working on latitude and longitude
but it is not working on time
when i give slice to time it giving error

```pytb
Traceback (most recent call last):
  File "/home/user/Documents/IGCAR/rain/rain.py", line 36, in <module>
     rainnc.isel(time=slice(0,8),lev_2=0).plot.pcolormesh(cmap='jet',vmax=100,vmin=0,add_colorbar=True);
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/plot/plot.py", line 1302, in plotmethod
    return newplotfunc(**allargs)
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/plot/plot.py", line 1132, in newplotfunc
    xlab, ylab = _infer_xy_labels(
  File "/home/user/anaconda3/lib/python3.9/site-packages/xarray/plot/utils.py", line 390, in _infer_xy_labels
    raise ValueError("DataArray must be 2d")
ValueError: DataArray must be 2d
```

**Reply by keewis** [reply]
to answer that we need more information. Can you post the output of `print(rainnc)`?

**Comment by haritha1022** [comment]
yeah sure
```
 Dimensions:  (time: 9, lon: 378, lat: 351, lev_2: 1)
Coordinates:
  * time     (time) datetime64[ns] 2021-12-30 2021-12-30T03:00:00 ... 2021-12-31
  * lon      (lon) float64 77.43 77.45 77.46 77.47 ... 81.32 81.33 81.34 81.35
  * lat      (lat) float64 11.14 11.15 11.16 11.17 ... 14.65 14.66 14.67 14.68
  * lev_2    (lev_2) float64 1e+03
Data variables:
    rainnc   (time, lev_2, lat, lon) float32 ...
```

**Reply by keewis** [reply]
The error is raised because `pcolormesh` needs a 2D array but the slice returns a 3D array (`(time, lat, lon)`). To fix that, you can try plotting each time separately by passing `col="time"`. If that's not what you wanted, can you describe what you wanted to do?

**Reply by earthtech23** [reply]
Hello @keewis 

I have a similar error to the one referred to the thread. I am already using col="time" to plot time series of satellite images (sentinel-2) and stills having the error ValueError: DataArray must be 2d

Any other ideas?

**Comment by haritha1022** [comment]
actually i want add the rainnc value from time step 0 to 8

**Comment by haritha1022** [comment]
i want to plot total rainnc data in a single plot
i want add the rainnc timesteps

**Comment by keewis** [comment, marked as answer]
then you're looking for
```python
rainnc.isel(time=slice(0, 8), lev_2=0).sum(dim="time")
```
which sums up the 8 time steps between the 0th and 7th (`slice`s are right-exclusive)

**Comment by haritha1022** [comment]
actually there are 9 time steps

**Reply by keewis** [reply]
then you have to use `slice(0, 9)`

**Comment by haritha1022** [comment]
I will try sir
thank you sir

**Comment by haritha1022** [comment]
Thanks a lot.
it works sir
